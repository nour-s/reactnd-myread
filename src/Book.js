import React from "react";
import PropTypes from "prop-types";

class Book extends React.Component {
	options = [
		{ value: "move", text: "Move to...", disabled: true },
		{ value: "currentlyReading", text: "Currently Reading" },
		{ value: "read", text: "Read" },
		{ value: "none", text: "None" }
	];

	onShelfChanged = select => {
		var shelfId = select.target.value;
		this.props.onMoveBook(this.props.book, shelfId);
	};

	render() {
		var book = this.props.book;
		return (
			<div className="book">
				<div className="book-top">
					<div
						className="book-cover"
						style={{
							width: 128,
							height: 193,
							backgroundImage: `url(${
								(book.imageLinks || { thumbnail: "" }).thumbnail
							})`
						}}
					/>
					<div className="book-shelf-changer">
						<select
							value={book.shelf || "none"}
							onChange={this.onShelfChanged}
						>
							{this.options.map(o => (
								<option
									key={o.value}
									value={o.value}
									disabled={o.disabled || false}
								>
									{o.text}
								</option>
							))}
						</select>
					</div>
				</div>
				<div className="book-title">{book.title}</div>
				{book.authors && (
					<div className="book-authors">
						{book.authors.reduce(
							(all, author) => `${all} - ${author}`
						)}
					</div>
				)}
			</div>
		);
	}
}

Book.propTypes = {
	book: PropTypes.shape({
		title: PropTypes.string,
		shelf: PropTypes.string,
		authors: PropTypes.array,
		imageLinks: PropTypes.shape({
			thumbnail: PropTypes.string
		})
	})
};
export default Book;
