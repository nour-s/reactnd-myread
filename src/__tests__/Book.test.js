import React from "react";
import renderer from "react-test-renderer";
import Book from "../Book";
import { mount } from "enzyme";

describe("Book test", () => {
	let rawComponent;
	let onMoveBook;
	let book = {
		title: "a book",
		shelf: "a shelf",
		authors: ["author 1", "author 2"],
		imageLinks: { thumbnail: "url" }
	};

	beforeEach(() => {
		onMoveBook = jest.fn();
		rawComponent = <Book onMoveBook={onMoveBook} book={book} />;
	});

	it("Book should match snapshot", () => {
		let component = renderer.create(rawComponent);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it("Book - move book to a shelf should work", () => {
		var wrapper = React.createElement(() => rawComponent);
		// I don't know why mount(rawComponent) doesn't work, and I have to do a work-around
		let component = mount(wrapper);
		const select = component.find("select").first();
		expect(select).toBeDefined();
		select.simulate("change", { target: { value: "read" } });
		expect(onMoveBook).toBeCalledWith(book, "read");
	});
});
