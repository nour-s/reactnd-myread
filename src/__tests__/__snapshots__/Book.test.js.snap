// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book test Book should match snapshot 1`] = `
<div
  className="book"
>
  <div
    className="book-top"
  >
    <div
      className="book-cover"
      style={
        Object {
          "backgroundImage": "url(url)",
          "height": 193,
          "width": 128,
        }
      }
    />
    <div
      className="book-shelf-changer"
    >
      <select
        onChange={[Function]}
        value="a shelf"
      >
        <option
          disabled={true}
          value="move"
        >
          Move to...
        </option>
        <option
          disabled={false}
          value="currentlyReading"
        >
          Currently Reading
        </option>
        <option
          disabled={false}
          value="read"
        >
          Read
        </option>
        <option
          disabled={false}
          value="none"
        >
          None
        </option>
      </select>
    </div>
  </div>
  <div
    className="book-title"
  >
    a book
  </div>
  <div
    className="book-authors"
  >
    author 1 - author 2
  </div>
</div>
`;
